File: src/main/java/com/github/nsc/de/compiler/interpreter/Scope.java
package com.github.nsc.de.compiler.interpreter;

/**
 * A scope to keep all the variables inside
 *
 * @author Nicolas Schmidt
 *
 * @see VariableList
 */
public class Scope {



    // *******************************
    // fields

    /**
     * The variables of the {@link Scope}
     */
    private final VariableList variables;

    /**
     * The parent {@link Scope} (or null if the {@link Scope} has no parent)
     */
    private final Scope parent;



    // *******************************
    // constructors

    /**
     * Constructor for {@link Scope}
     *
     * @param parent the parent of the scope (value for {@link #parent})
     * @param variables the variables of the scope (value for {@link #variables})
     *
     * @author Nicolas Schmidt
     */
    public Scope(Scope parent, VariableList variables) {
        // just set the fields
        this.parent = parent;
        this.variables = variables;
    }

    /**
     * Constructor for {@link Scope}
     *
     * @param parent the parent of the scope (value for {@link #parent})
     *
     * @author Nicolas Schmidt
     */
    public Scope(Scope parent) {
        // Use other constructor with an empty VariableList
        this(parent, new VariableList());
    }



    // *******************************
    // getters

    /**
     * Getter for {@link #parent} (the parent-scope)
     *
     * @return the parent-scope (this.{@link #parent})
     *
     * @author Nicolas Schmidt
     */
    public Scope getParent() {
        return parent;
    }

    /**
     * Getter for {@link #variables} (the scope's variables)
     *
     * @return the scope's variables (this.{@link #variables})
     *
     * @author Nicolas Schmidt
     */
    public VariableList getScopeVariables() {
        return variables;
    }

    /**
     * Getter for all scope variables (including all the variables of the parents)
     *
     * @return the scope's variables (this.{@link #variables})
     *
     * @author Nicolas Schmidt
     */
    public VariableList getVariables() {
        return this.parent != null ? this.parent.getVariables().concat(variables) : variables;
    }

    /**
     * Copies the scope
     *
     * @return the copy of the scope
     *
     * @author Nicolas Schmidt
     */
    public Scope copy() {
        return new Scope(this.parent, this.getScopeVariables().copy());
    }
}

File: shake/compiler/interpreter/src/commonMain/kotlin/io/github/shakelang/shake/interpreter/Scope.kt
package io.github.shakelang.shake.interpreter

import io.github.shakelang.shake.interpreter.values.InterpreterValue

/**
 * A scope to keep all the variables inside
 *
 *
 * @see VariableList
 */
class Scope


/**
 * Constructor for [Scope]
 *
 * @param parent the parent of the scope (value for [parent])
 * @param scopeVariables the variables of the scope (value for [scopeVariables])
 * @param interpreter the [interpreter] of the scope
 */
    (

    /**
     * The parent [Scope] (or null if the [Scope] has no parent)
     */
    val parent: Scope?,

    /**
     * The variables of the [Scope]
     */
    val scopeVariables: VariableList,

    /**
     * The interpreter of the [Scope]
     */
    val interpreter: Interpreter

) {

    // *******************************
    // fields

    /**
     * Getter for the [.returnValue] field
     *
     * @return the [.returnValue]
     */
    var returnValue: InterpreterValue? = null

    /**
     * Constructor for [Scope]
     *
     * @param parent the parent of the scope (value for [.parent])
     */
    constructor(parent: Scope?, interpreter: Interpreter) : this(parent, VariableList(), interpreter)


    // *******************************
    // getters

    /**
     * Getter for all scope variables (including all the variables of the parents)
     *
     * @return the scope's variables (this.[.variables])
     */
    fun getVariables(): VariableList {
        return if (this.parent != null) this.parent.getVariables().concat(this.scopeVariables) else this.scopeVariables
    }

    /**
     * Copies the scope
     *
     * @return the copy of the scope
     */
    fun copy(): Scope {
        return Scope(this.parent, this.scopeVariables.copy(), interpreter)
    }

    fun reset() {
        this.getVariables().reset()
    }
    // *******************************
    // constructors
}